src/main/CommandHandler.cpp

```diff
@@ -678,6 +712,79 @@ CommandHandler::unban(std::string const& params, std::string& retStr)
     }
 }
 
+void
+CommandHandler::upgrades(std::string const& params, std::string& retStr)
+{
+    std::map<std::string, std::string> retMap;
+    http::server::server::parseParams(params, retMap);
+    auto s = retMap["mode"];
+    if (s.empty())
+    {
+        retStr = "mode required";
+        return;
+    }
+    if (s == "get")
+    {
+        retStr = mApp.getHerder().getUpgrades();
+    }
+    else if (s == "set")
+    {
+        Upgrades::UpgradeParameters p;
+
+        auto upgradeTime = retMap["upgradetime"];
+
+        std::istringstream is(upgradeTime);
+        std::tm tm;
+        is >> std::get_time(&tm, "%Y-%m-%dT%H:%M:%SZ");
+        if (is.fail())
+        {
+            retStr =
+                fmt::format("could not parse upgradetime: '{}'", upgradeTime);
+            return;
+        }
+        tm.tm_isdst = false;
+        p.mUpgradeTime = VirtualClock::tmToPoint(tm);
+
+        auto addParam = [&](std::string const& name,
+                            stellar::optional<uint32>& f) {
+            uint32 v;
+            bool updated;
+            if (!parseNumParam(retMap, name, v, retStr,
```

Comment:
I start to think we should use exceptions inside parseNumParam and just return optional<uint32> from it.
Now its API is very unnatural - it has 3 return values - bool, uint32 and another bool.

Reply:
Yeah that would make sense, especially that now I made the exception handler global. I created an issue for this